### ISSUE TITLE
Potential Issue w/ AGEM with multi_head plugin

### ISSUE THREAD

**Issue by arnavmdas** [ContinualAI/avalanche#468]
**Description**
When looking at the code for agem.py (in the before_training_iteration function),  it appears that the reference gradient is computed w.r.t the head of the current task as opposed to being computed w.r.t to the head of corresponding the the task to which each sample belongs.  It does not seem like the before_training_iteration checks whether or not the multi_head plugin is present.

In this code snippet (from the agem before_training_iteration function), the task descriptor for each member of x_ref is not accounted for and it doesn't seem like there is a check  for the multi_head plugin anywhere:

![image](https://user-images.githubusercontent.com/25560726/113047299-c142bc00-9166-11eb-9636-b45b9372fe07.png)

I think that this is inconsistent with the original AGEM implementation, and may therefore lead to a degradation in performance.

**Comment by AndreaCossu** [comment]
Hi @arnavmdas :smile: In my opinion, the multi-head is orthogonal to A-GEM: A-GEM is a strategy which projects the gradients. It can work with both single and multi-head plugin (with different final performances, of course). If you are trying to use a multi-head setting, you can leverage the Avalanche MultiHead plugin in combination with A-GEM. With Multi-head plugin active at training time, there should be no need to check its presence in A-GEM: the gradient of inactive units should be zero (@AntonioCarta can you confirm this?).

As far as task descriptors are concerned you are right, at the moment our version of A-GEM does not allow for the use of task descriptors. If you feel like to, you can try to modify it and submit a PR :smile: I guess that for most common benchmarks, a simple one-hot task vector should be enough. Feel free to open an issue on that!

**Comment by arnavmdas** [comment]
Thanks for your response!

I am currently using the multihead plugin along with the AGEM plugin but we are unfortunately unable to replicate the original CIFAR100 results. 

Currently, we are using 20 experiences on CIFAR100 so each experience consists of 5 different labels (which is also the number of output nodes in the model when the multihead plugin is used). Because of this, the labels in yref all take values between 0 and 5, rather than between 0 and 100. The gradient of the inactive units will only be zero if some sort of mask was used for the multihead implementation which doesn't seem to be the case. Furthermore, since the task descriptors for each sample are not tracked there is no way of knowing which head to use when computing the gradient, so the reference gradient may not be correct.

**Comment by AntonioCarta** [comment]
Thank you @arnavmdas! You are correct. If the wrong head is active (as it is currently), the output will be wrong, and therefore the gradients. This happens because the correct head is set inside a callback, which is not called again by A-GEM.

I see a couple of ways to solve this:
- force strategies and plugins to call `strategy.forward()` instead of `model()`. This is error-prone because it's something unexpected.
- add forward hooks to the model to set the correct head. This is maybe the best solution and transparent to the user.
- refactor the multi-head plugin to become a `nn.Module` which switches heads internally. This is almost transparent to the user (requires changing your model) but requires to pass the task label as an additional argument to the model, which may not work with pretrained/predefined architectures.

Let me think about this. Thanks again for reporting this issue.

If you have a minimal script that I can run to check this problem it would be helpful.

**Comment by bhattg** [comment]
Hello, is this issue now resolved?

**Comment by AntonioCarta** [comment]
The multi-head is fixed. However, there may be some other issues with A-GEM because @AndreaCossu was not able to reproduce the paper's experiments. We are working on them.

**Comment by vlomonaco** [comment]
On this I'd like to mention here for future reference that we are now starting a larger effort to reproduce papers results in this avalanche spin-off project: https://github.com/ContinualAI/reproducible-continual-learning

**Comment by bhattg** [comment]
Thanks for the information!!!
